### PR TITLE
Add possibility to edit min and max of intensity range

### LINKED ISSUE #4526: Expose min/max for float layers
With float layers the histogram is necessary for any viewing of the dataset. Sometimes the sampling does not yield sensible results. For these datasets, it would be nice to specify min and max with an input field. Please note that the range of floats can be both very small (e.g. 4.8e-15) and very large (e.g. 3.2e+12).

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
This PR adds the possibility for the user to edit the minimum and maximum possible value for the intensity range. 
The use case is described in #4526.

### URL of deployed dev instance (used for testing):
- https://addminmaxediting.webknossos.xyz

### Steps to test:
- Open a dataset (view mode is enough)
- Toggle the edit function for min and max. The button is on the left of the reload-layer button.
- enter some desired values
- the histogram and the `intensityRange` slider's range should adjust to the entered values.
- then adjust the `intensityRange` and you should be able to see something again :)

- please try breaking the feature by entering some random and uncommon values

### Issues:
- fixes #4526

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- ~~[ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable~~
- ~~[ ] Updated [documentation](../blob/master/docs) if applicable~~
- ~~[ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes~~
- ~~[ ] Needs datastore update after deployment~~
- [x] Ready for review

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
Thanks a lot for finding those bugs :bug:. Hopefully, I fixed them all. Could you please check that again?

>I'm not sure about the scientific notation feature. Try entering a large max value like 400123 by hand - you end up with 4.0012e+43 which is not what you wanted to type. It's confusing that the value changes while you type. If we still want the scientific notation feature to avoid overly long numbers, I would suggest to add a mechanism to confirm the entered number by clicking on a button. Then, the number could change to scientific notation. However, I don't think this is very important.


Thus we should not support displying in scientific notation? Is that what you meant?

**Comment by daniel-wer** [comment]
> Thus we should not support displying in scientific notation? Is that what you meant?

Yes, I don't think it is very important, and I don't see an easy way to implement it with good UX, so I wouldn't add it for now.

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 1]
```diff
@@ -138,8 +138,10 @@ class Histogram extends React.PureComponent<HistogramProps> {
     }
     ctx.stroke();
     ctx.closePath();
-    activeRegion.lineTo(((intensityRangeMax - minRange) / fullLength) * canvasWidth, 0);
-    activeRegion.lineTo(((intensityRangeMin - minRange) / fullLength) * canvasWidth, 0);
+    const activeRegionRightLimit = Math.min(histogramMax, intensityRangeMax);
+    const activeRegionLeftLimit = Math.max(histogramMin, intensityRangeMin);
+    activeRegion.lineTo(((activeRegionRightLimit - minRange) / fullLength) * canvasWidth, 0);
+    activeRegion.lineTo(((activeRegionLeftLimit - minRange) / fullLength) * canvasWidth, 0);
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:144]: @MichaelBuessemeyer My assumption would be that `histogramMax` is always <= `intensityRangeMax` and therefore, it should be sufficient to simply use `histogramMax` in the `activeRegion.lineTo` call. Is that assumption correct?

Comment (daniel-wer) [reply frontend/javascripts/oxalis/view/settings/histogram_view.js:144]: Maybe I was confusing the `histogramMax` and `intensityRangeMax`. Hmm, could you distinguish what each of those is again for me? Maybe we can find more descriptive names that make the difference more clear.

Comment (MichaelBuessemeyer) [reply frontend/javascripts/oxalis/view/settings/histogram_view.js:144]: I think you really confused both of them:
- `histogramMax` is the maximum color value that is included in the histogram data send by the backend.
- `intensityRangeMax` is the maximum color value of the displayed color range and is equal to the higher value of the slide. What's new now is that this value can be greater than `histogramMax` as you can see it in your example:
max_range_float

To be honest I think the names are pretty good as the `intensityRange` is the range given by the slider and thus `intensityRangeMax` as the upper limit of that range is not far-fetched (in my opinion) -> Feel free to argue about that.

The same goes for `histogramMax` as this is the highest color value included in the histogram data.

Comment (daniel-wer) [reply frontend/javascripts/oxalis/view/settings/histogram_view.js:144]: Ah understood, that makes sense, sorry for the confusion :)

**Comment by MichaelBuessemeyer** [comment]
> The behavior when entering an entirely new number is kind of weird. For example, the max value is set to 100, now you enter 255 instead. When entering the 2, the max slider of the histogram is automatically set to 2, and when entering the 55 afterwards, the max slider remains at 2. However, the max slider was set to 100 before and it's a little unexpected that this changed (although I understand why it happened). A possible solution for this could be to debounce the onChange of the input field so that the histogram only changes after a second of no change or so. What do you think?


I tried to tackle this by only updating when the events `onPressEnter` and `onBlur` are executed but apparently `onPressEnter` is not available for NumberInput in our antd version. :confounded: Thus I choose to just debounce the update. I tested this and it feels pretty smooth :+1: 

>  [Low Pri] Maybe it's an easy fix, if not nevermind: The two vertical side lines of the active area are not vertical, but slightly angled to the middle. To see this enter 100 and 102 for min and max and look at the active area.

I cannot find the cause for this :confused: 

@daniel-wer Could you please check the small updates I just pushed?

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 2]
```diff
@@ -32,6 +32,11 @@ type HistogramProps = {
   ) => void,
 };
 
+type HistorgramState = {
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:35]:
```suggestion
type HistogramState = {
```

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 3]
```diff
@@ -45,9 +50,25 @@ export function isHistogramSupported(elementClass: ElementClass): boolean {
...
+  const { min, max, data } = props;
+  if (min != null && max != null) {
+    return { min, max };
+  }
+  const dataMin = Math.min(...data.map(({ min: minOfHistPart }) => minOfHistPart));
+  const dataMax = Math.max(...data.map(({ max: maxOfHistPart }) => maxOfHistPart));
+  return { min: dataMin, max: dataMax };
+}
+
+class Histogram extends React.PureComponent<HistogramProps, HistorgramState> {
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:63]:
```suggestion
class Histogram extends React.PureComponent<HistogramProps, HistogramState> {
```

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 4]
```diff
@@ -183,7 +220,8 @@ class Histogram extends React.PureComponent<HistogramProps> {
       layerName,
       onChangeLayer,
     } = this.props;
-    const { min: minRange, max: maxRange } = this.getMinAndMax();
+    const { currentMin, currentMax } = this.state;
+    const { min: minRange, max: maxRange } = getMinAndMax(this.props);
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:223]: Looks like `this.getMinAndMax` can be removed now :)

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 5]
```diff
@@ -132,9 +197,41 @@ class Histogram extends React.PureComponent<HistogramProps> {
     }
   };
 
+  tipFormatter = (value: number) =>
+    value > 10000 ? value.toExponential() : roundTo(value, 2).toString();
+
+  // eslint-disable-next-line react/sort-comp
+  updateMinimumDebounced = _.debounce(
+    (value, layerName) => this.props.onChangeLayer(layerName, "min", value),
+    1000,
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:206]:
```suggestion
    500,
```

I would suggest to half the debounce time to make it feel a little more snappy

File: frontend/javascripts/oxalis/view/settings/histogram_view.js [hunk 6]
```diff
@@ -132,9 +197,41 @@ class Histogram extends React.PureComponent<HistogramProps> {
...
+  );
+
+  updateMaximumDebounced = _.debounce(
+    (value, layerName) => this.props.onChangeLayer(layerName, "max", value),
+    1000,
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/settings/histogram_view.js:211]:
```suggestion
    500,
```

**Comment by MichaelBuessemeyer** [comment]
@daniel-wer Thanks for the suggestions. I just applied them.

**Pull request merged**